### PR TITLE
Tests for Schema Compare utils file

### LINKED ISSUE #8738: Create Unit test for Schema Compare- SQL Login feature

### PR DESCRIPTION
**Pull request by SakshiS-harma**
This PR introduces ApiWrapper in SchemaCompare for appropriate unit testing.
Added tests for utils.ts (coverage of the file is ~92% with these new tests. Originally ~41%).

This PR also fixes #8738

### CODE REVIEW

File: extensions/schema-compare/src/test/schemaCompare.test.ts [hunk 1]
```diff
@@ -79,22 +84,22 @@ describe('SchemaCompareResult.start', function (): void {
...
+		let result = new SchemaCompareMainWindow(testContext.apiWrapper.object, sc, mockExtensionContext.object);
+		await result.start({connectionProfile: mockIConnectionProfile});
 		let promise = new Promise(resolve => setTimeout(resolve, 5000)); // to ensure comparison result view is initialized
 		await promise;
 
 		should.notEqual(result.sourceEndpointInfo, undefined);
 		should.equal(result.sourceEndpointInfo.endpointType, mssql.SchemaCompareEndpointType.Database);
-		should.equal(result.sourceEndpointInfo.serverName, mockConnectionProfile.serverName);
-		should.equal(result.sourceEndpointInfo.databaseName, mockConnectionProfile.databaseName);
+		should.equal(result.sourceEndpointInfo.serverName, mockIConnectionProfile.serverName);
```
Comment (udeeshagautam) [inline extensions/schema-compare/src/test/schemaCompare.test.ts:94]: ah goo that we already have a test to create SCMainWindow with a Source Endpoint - can we add a test to update/change the endpoint too?

Also in this test there should be a check that target endpoint is empty/undefiled - right?

File: extensions/schema-compare/src/test/schemaCompare.test.ts [hunk 2]
```diff
@@ -79,22 +84,22 @@ describe('SchemaCompareResult.start', function (): void {
...
+		should.equal(result.sourceEndpointInfo.databaseName, mockIConnectionProfile.databaseName);
 		should.equal(result.targetEndpointInfo, undefined);
 	});
 
 	it('Should start with the source as dacpac.', async function (): Promise<void> {
 		let sc = new SchemaCompareTestService();
 
-		let result = new SchemaCompareMainWindow(sc, mockExtensionContext.object);
+		let result = new SchemaCompareMainWindow(testContext.apiWrapper.object, sc, mockExtensionContext.object);
 		const dacpacPath = 'C:\\users\\test\\test.dacpac';
```
Comment (udeeshagautam) [inline extensions/schema-compare/src/test/schemaCompare.test.ts:103]: seems weird to have the windows path (if this is expected can you add a comment to explain this please) - I know not a part of your PR

File: extensions/schema-compare/src/test/testContext.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
```
Comment (udeeshagautam) [inline extensions/schema-compare/src/test/testContext.ts:5]: should this file be names testcontext.test.ts - can you check how does a file not names *.test but inside test folder impact Coverage?

File: extensions/schema-compare/src/test/utils.test.ts [hunk 4]
```diff
@@ -35,10 +41,14 @@ describe('utils: Tests to verify getEndpointName', function (): void {
 	});
 });
 
-describe('utils: Tests to verify verifyConnectionAndGetOwnerUri', function (): void {
+describe('utils: Basic tests to verify verifyConnectionAndGetOwnerUri', function (): void {
+	before(async function (): Promise<void> {
+		testContext = createContext();
+	});
+
 	it('Should return undefined for endpoint as dacpac', async function (): Promise<void> {
```
Comment (udeeshagautam) [inline extensions/schema-compare/src/test/utils.test.ts:49]: test for what should it return when call with connection rather than dacpac?

File: extensions/schema-compare/src/test/utils.test.ts [hunk 5]
```diff
@@ -48,8 +58,58 @@ describe('utils: Tests to verify verifyConnectionAndGetOwnerUri', function (): v
...
+		testDatabaseEndpoint.connectionDetails = {...mockConnectionInfo};
+
+		testContext.apiWrapper.setup(x => x.connect(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => { return Promise.resolve(connection); });
+		testContext.apiWrapper.setup(x => x.getUriForConnection(TypeMoq.It.isAny())).returns(() => { return Promise.resolve(undefined); });
+		testContext.apiWrapper.setup(x => x.getConnections(TypeMoq.It.isAny())).returns(() => { return Promise.resolve(getConnectionsResults); });
+		testContext.apiWrapper.setup(x => x.showWarningMessage(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => { return Promise.resolve(loc.YesButtonText); });
+		testContext.apiWrapper.setup(x => x.openConnectionDialog(TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => { return Promise.resolve(undefined); });
+
+		await shouldThrowSpecificError(async () => await verifyConnectionAndGetOwnerUri(testDatabaseEndpoint, 'test', testContext.apiWrapper.object), connection.errorMessage);
+	});
```
Comment (kisantia) [inline extensions/schema-compare/src/test/utils.test.ts:114]: can there also be a positive test for when verifyConnectionAndGetOwnerUri() returns successfully?

File: extensions/schema-compare/src/test/utils.test.ts [hunk 6]
```diff
@@ -48,8 +58,58 @@ describe('utils: Tests to verify verifyConnectionAndGetOwnerUri', function (): v
...
+
+		testContext.apiWrapper.setup(x => x.connect(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => { return Promise.resolve(connection); });
+		testContext.apiWrapper.setup(x => x.getUriForConnection(TypeMoq.It.isAny())).returns(() => { return Promise.resolve(undefined); });
+		testContext.apiWrapper.setup(x => x.getConnections(TypeMoq.It.isAny())).returns(() => { return Promise.resolve(getConnectionsResults); });
+		testContext.apiWrapper.setup(x => x.showErrorMessage(TypeMoq.It.isAny())).returns((s) => { throw new Error(s); });
+
+		await shouldThrowSpecificError(async () => await verifyConnectionAndGetOwnerUri(testDatabaseEndpoint, 'test', testContext.apiWrapper.object), connection.errorMessage);
+	});
+
+	it('Should not throw an error for login failure with openConnectionDialog', async function (): Promise<void> {
```
Comment (udeeshagautam) [inline extensions/schema-compare/src/test/utils.test.ts:101]: should or should not? line 113 seems like it does throw.

File: extensions/schema-compare/src/test/testContext.ts (hunk 3 above)
Comment (udeeshagautam) [reply extensions/schema-compare/src/test/testContext.ts:5]: @kisantia do you have any idea here around file names playing any role in CC?

Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/testContext.ts:5]: Anything that is part of test folder, isn't tested for Coverage, so the naming here doesn't matter. 
Moreover, *.test.ts files are the ones that actually have some tests that are run. So this wouldn't qualify.

Comment (udeeshagautam) [reply extensions/schema-compare/src/test/testContext.ts:5]: sounds good. thanks!

File: extensions/schema-compare/src/test/utils.test.ts (hunk 6 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/utils.test.ts:101]: My bad, it should be "should". I will remove the "not".

File: extensions/schema-compare/src/test/schemaCompare.test.ts (hunk 1 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompare.test.ts:94]: I will add the suggested test in another round of test PR.
As for second, as discussed, the check already exists.

File: extensions/schema-compare/src/test/schemaCompare.test.ts (hunk 2 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompare.test.ts:103]: It is just a dummy file path. Updated it to 'test.dacpac'- so won't be platform specific.

File: extensions/schema-compare/src/test/utils.test.ts (hunk 5 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/utils.test.ts:114]: Added one more test here, for the positive scenario.
